OrbitCore/CallstackData.cpp

```diff
@@ -175,3 +176,58 @@ std::shared_ptr<CallStack> CallstackData::GetCallstackPtr(CallstackID callstack_
   }
   return nullptr;
 }
+
+void CallstackData::FilterCallstackEventsBasedOnMajorityStart() {
+  std::lock_guard lock(mutex_);
+  uint32_t count_before_filtering = GetCallstackEventsCount();
+
+  for (auto& tid_and_events : callstack_events_by_tid_) {
+    std::map<uint64_t, CallstackEvent>& callstack_events = tid_and_events.second;
+
+    // Count the number of occurrences of each outer frame for this thread.
+    absl::flat_hash_map<uint64_t, uint64_t> count_by_outer_frame;
+    for (const auto& time_and_event : callstack_events) {
+      const CallstackEvent& event = time_and_event.second;
+      const std::vector<uint64_t> frames =
+          unique_callstacks_.at(event.callstack_hash())->GetFrames();
+      if (frames.empty()) {
+        continue;
+      }
+      uint64_t outer_frame = *frames.rbegin();
+      ++count_by_outer_frame[outer_frame];
+    }
+
+    // Find the outer frame with the most occurrences.
+    uint64_t majority_outer_frame = 0;
+    uint64_t majority_outer_frame_count = 0;
+    for (const auto& outer_frame_and_count : count_by_outer_frame) {
+      if (outer_frame_and_count.second > majority_outer_frame_count) {
+        majority_outer_frame = outer_frame_and_count.first;
+        majority_outer_frame_count = outer_frame_and_count.second;
+      }
+    }
+    if (majority_outer_frame_count == 0) {
+      continue;
+    }
```

Comment:
maybe also check for supermajority here? majority_outer_frame_count > 0.6*number of callstacks in the thread.

Just a safeguard for distributions like (21, 20, 20, 19, 20) where 21 is winning but something this picture does not look right.

Reply:
What do you think the action should be in that case though? Should the filtering just not happen? Should all events be removed (that's a bit extreme)? Should the filtering stay the same but at least an error logged?

Reply:
I think filtering should not be happening at this point - maybe print a warning/error message.

Reply:
Done, let me know what you think.